File: constructs/java/src/test/java/com/k1cka5h/infra/ConstructsTest.java
package com.k1cka5h.infra;

import com.hashicorp.cdktf.App;
import com.hashicorp.cdktf.Testing;
import com.hashicorp.cdktf.TerraformStack;
import com.k1cka5h.infra.constructs.AksConstruct;
import com.k1cka5h.infra.constructs.DatabaseConstruct;
import com.k1cka5h.infra.constructs.NetworkConstruct;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Construct unit tests — Java
 * ============================
 * Synthesizes each construct to JSON and asserts the module call is wired
 * correctly. Does not run Terraform or touch Azure.
 *
 * Run:
 *   mvn test
 */
class ConstructsTest {

    // ── helpers ────────────────────────────────────────────────────────────────

    /** Plain TerraformStack for constructs that don't need a full BaseAzureStack. */
    private static TerraformStack plainStack() {
        App app = Testing.app();
        return new TerraformStack(app, "test-stack");
    }

    /** Synthesize a stack to JSON (returned as String). */
    private static String synth(TerraformStack stack) {
        return Testing.synthScope(stack);
    }


    // ── BaseAzureStack ─────────────────────────────────────────────────────────

    @Test
    void baseStack_setsCorrectStateKey() {
        App app = Testing.app();
        BaseAzureStack stack = new BaseAzureStack(app, "base",
                BaseAzureStack.BaseAzureStackProps.builder()
                        .project("proj")
                        .environment("dev")
                        .build());

        String json = Testing.synthScope(stack);
        assertTrue(json.contains("\"proj/dev/terraform.tfstate\""),
                "state key should be proj/dev/terraform.tfstate");
    }

    @Test
    void baseStack_defaultLocationIsEastUs() {
        App app = Testing.app();
        BaseAzureStack stack = new BaseAzureStack(app, "base2",
                BaseAzureStack.BaseAzureStackProps.builder()
                        .project("proj")
                        .environment("staging")
                        .build());

        assertEquals("eastus", stack.getLocation());
    }

    @Test
    void baseStack_throwsOnInvalidEnvironment() {
        App app = Testing.app();
        assertThrows(IllegalArgumentException.class, () ->
                new BaseAzureStack(app, "bad",
                        BaseAzureStack.BaseAzureStackProps.builder()
                                .project("proj")
                                .environment("uat")
                                .build()));
    }


    // ── NetworkConstruct ───────────────────────────────────────────────────────

    @Test
    void network_moduleSourceIsPinned() {
        TerraformStack stack = plainStack();
        new NetworkConstruct(stack, "net",
                NetworkConstruct.NetworkConstructProps.builder()
                        .project("myapp")
                        .environment("dev")
                        .resourceGroup("myapp-dev-rg")
                        .location("eastus")
                        .addressSpace(List.of("10.0.0.0/16"))
                        .build());

        String json = synth(stack);
        assertTrue(json.contains("ref=v"),
                "module source should pin a ref tag");
        assertTrue(json.contains("modules/networking"),
                "module source should reference modules/networking");
    }

    @Test
    void network_requiredTagsInjected() {
        TerraformStack stack = plainStack();
        new NetworkConstruct(stack, "net",
                NetworkConstruct.NetworkConstructProps.builder()
                        .project("myapp")
                        .environment("staging")
                        .resourceGroup("myapp-staging-rg")
                        .location("eastus")
                        .addressSpace(List.of("10.0.0.0/16"))
                        .build());

        String json = synth(stack);
        assertTrue(json.contains("\"managed_by\""),   "tags should contain managed_by");
        assertTrue(json.contains("\"terraform\""),    "managed_by value should be terraform");
        assertTrue(json.contains("\"myapp\""),         "tags should contain project=myapp");
        assertTrue(json.contains("\"staging\""),       "tags should contain environment=staging");
    }

    @Test
    void network_subnetDelegationSerialized() {
        TerraformStack stack = plainStack();
        new NetworkConstruct(stack, "net",
                NetworkConstruct.NetworkConstructProps.builder()
                        .project("myapp")
                        .environment("dev")
                        .resourceGroup("myapp-dev-rg")
                        .location("eastus")
                        .addressSpace(List.of("10.0.0.0/16"))
                        .subnets(Map.of(
                                "db", NetworkConstruct.SubnetConfig.builder()
                                        .addressPrefix("10.0.8.0/24")
                                        .delegation(NetworkConstruct.SubnetDelegation.builder()
                                                .name("postgres")
                                                .service("Microsoft.DBforPostgreSQL/flexibleServers")
                                                .actions(List.of("Microsoft.Network/virtualNetworks/subnets/join/action"))
                                                .build())
                                        .build()))
                        .build());

        String json = synth(stack);
        assertTrue(json.contains("Microsoft.DBforPostgreSQL/flexibleServers"),
                "subnet delegation service should appear in synth output");
    }


    // ── DatabaseConstruct ──────────────────────────────────────────────────────

    @Test
    void database_moduleSourceIsPinned() {
        TerraformStack stack = plainStack();
        new DatabaseConstruct(stack, "db",
                DatabaseConstruct.DatabaseConstructProps.builder()
                        .project("myapp")
                        .environment("prod")
                        .resourceGroup("myapp-prod-rg")
                        .location("eastus")
                        .subnetId("/subscriptions/x/subnets/db")
                        .dnsZoneId("/subscriptions/x/privateDnsZones/postgres.database.azure.com")
                        .adminPassword("Hunter2!")
                        .build());

        String json = synth(stack);
        assertTrue(json.contains("ref=v"),               "module source should pin a ref tag");
        assertTrue(json.contains("modules/database/postgres"), "source should reference postgres module");
    }

    @Test
    void database_haEnabledSetsZoneRedundant() {
        TerraformStack stack = plainStack();
        new DatabaseConstruct(stack, "db",
                DatabaseConstruct.DatabaseConstructProps.builder()
                        .project("myapp")
                        .environment("prod")
                        .resourceGroup("myapp-prod-rg")
                        .location("eastus")
                        .subnetId("/subscriptions/x/subnets/db")
                        .dnsZoneId("/subscriptions/x/privateDnsZones/postgres.database.azure.com")
                        .adminPassword("Hunter2!")
                        .config(DatabaseConstruct.PostgresConfig.builder()
                                .haEnabled(true)
                                .databases(List.of("appdb"))
                                .build())
                        .build());

        String json = synth(stack);
        assertTrue(json.contains("\"ZoneRedundant\""),
                "HA enabled should set high_availability_mode=ZoneRedundant");
    }

    @Test
    void database_haDisabledSetsDisabled() {
        TerraformStack stack = plainStack();
        new DatabaseConstruct(stack, "db",
                DatabaseConstruct.DatabaseConstructProps.builder()
                        .project("myapp")
                        .environment("dev")
                        .resourceGroup("myapp-dev-rg")
                        .location("eastus")
                        .subnetId("/subscriptions/x/subnets/db")
                        .dnsZoneId("/subscriptions/x/privateDnsZones/postgres.database.azure.com")
                        .adminPassword("Hunter2!")
                        .build());

        String json = synth(stack);
        assertTrue(json.contains("\"Disabled\""),
                "HA disabled should set high_availability_mode=Disabled");
    }

    @Test
    void database_adminPasswordForwarded() {
        TerraformStack stack = plainStack();
        new DatabaseConstruct(stack, "db",
                DatabaseConstruct.DatabaseConstructProps.builder()
                        .project("myapp")
                        .environment("dev")
                        .resourceGroup("myapp-dev-rg")
                        .location("eastus")
                        .subnetId("/subscriptions/x/subnets/db")
                        .dnsZoneId("/subscriptions/x/privateDnsZones/postgres.database.azure.com")
                        .adminPassword("MySecret123!")
                        .build());

        String json = synth(stack);
        assertTrue(json.contains("administrator_password"),
                "administrator_password variable should appear in module variables");
    }


    // ── AksConstruct ──────────────────────────────────────────────────────────

    @Test
    void aks_moduleSourceIsPinned() {
        TerraformStack stack = plainStack();
        new AksConstruct(stack, "aks",
                AksConstruct.AksConstructProps.builder()
                        .project("myapp")
                        .environment("dev")
                        .resourceGroup("myapp-dev-rg")
                        .location("eastus")
                        .subnetId("/subscriptions/x/subnets/aks")
                        .logWorkspaceId("/subscriptions/x/workspaces/logs")
                        .build());

        String json = synth(stack);
        assertTrue(json.contains("ref=v"),             "module source should pin a ref tag");
        assertTrue(json.contains("modules/compute/aks"), "source should reference aks module");
    }

    @Test
    void aks_additionalNodePoolForwarded() {
        TerraformStack stack = plainStack();
        new AksConstruct(stack, "aks",
                AksConstruct.AksConstructProps.builder()
                        .project("myapp")
                        .environment("staging")
                        .resourceGroup("myapp-staging-rg")
                        .location("eastus")
                        .subnetId("/subscriptions/x/subnets/aks")
                        .logWorkspaceId("/subscriptions/x/workspaces/logs")
                        .config(AksConstruct.AksConfig.builder()
                                .systemNodeCount(3)
                                .additionalNodePools(Map.of(
                                        "workers", AksConstruct.NodePoolConfig.builder()
                                                .vmSize("Standard_D8s_v3")
                                                .enableAutoScaling(true)
                                                .minCount(2)
                                                .maxCount(10)
                                                .build()))
                                .build())
                        .build());

        String json = synth(stack);
        assertTrue(json.contains("\"workers\""),         "worker pool name should appear in synth");
        assertTrue(json.contains("Standard_D8s_v3"),     "worker pool vm_size should appear in synth");
        assertTrue(json.contains("enable_auto_scaling"),  "auto scaling config should be present");
    }

    @Test
    void aks_requiredTagsInjected() {
        TerraformStack stack = plainStack();
        new AksConstruct(stack, "aks",
                AksConstruct.AksConstructProps.builder()
                        .project("myapp")
                        .environment("prod")
                        .resourceGroup("myapp-prod-rg")
                        .location("eastus")
                        .subnetId("/subscriptions/x/subnets/aks")
                        .logWorkspaceId("/subscriptions/x/workspaces/logs")
                        .build());

        String json = synth(stack);
        assertTrue(json.contains("\"managed_by\""), "tags should contain managed_by");
        assertTrue(json.contains("\"terraform\""),  "managed_by should be terraform");
        assertTrue(json.contains("\"prod\""),        "tags should contain environment=prod");
    }
}

File: constructs/cdktf/csharp/Nautilus.Infra.Tests/ConstructsTest.cs
using System.Text.Json;
using HashiCorp.Cdktf;
using Xunit;

namespace Nautilus.Infra.Tests;

/// <summary>
/// Construct unit tests — C#
/// ==========================
/// Synthesizes each construct to JSON and asserts the module call is wired
/// correctly. Does not run Terraform or touch Azure.
///
/// Run:
///   dotnet test Nautilus.Infra.Tests/
/// </summary>
public class ConstructsTest
{
    // ── helpers ────────────────────────────────────────────────────────────────

    private static TerraformStack PlainStack()
    {
        var app = Testing.App();
        return new TerraformStack(app, "test-stack");
    }

    private static JsonElement Synth(TerraformStack stack)
    {
        var json = Testing.SynthScope(stack);
        return JsonSerializer.Deserialize<JsonElement>(json);
    }

    private static JsonElement Module(JsonElement synth, string keyContains)
    {
        foreach (var prop in synth.GetProperty("module").EnumerateObject())
            if (prop.Name.Contains(keyContains))
                return prop.Value;
        throw new KeyNotFoundException($"No module key containing '{keyContains}'");
    }


    // ── BaseAzureStack ─────────────────────────────────────────────────────────

    [Fact]
    public void BaseStack_SetsCorrectStateKey()
    {
        var app   = Testing.App();
        var stack = new BaseAzureStack(app, "base",
            new BaseAzureStackProps("proj", "dev"));

        var synth = Synth(stack);
        var key   = synth
            .GetProperty("terraform")
            .GetProperty("backend")
            .GetProperty("azurerm")
            .GetProperty("key")
            .GetString();

        Assert.Equal("proj/dev/terraform.tfstate", key);
    }

    [Fact]
    public void BaseStack_DefaultLocationIsEastUs()
    {
        var app   = Testing.App();
        var stack = new BaseAzureStack(app, "base2",
            new BaseAzureStackProps("proj", "staging"));

        Assert.Equal("eastus", stack.Location);
    }

    [Fact]
    public void BaseStack_ThrowsOnInvalidEnvironment()
    {
        var app = Testing.App();
        Assert.Throws<ArgumentException>(() =>
            new BaseAzureStack(app, "bad",
                new BaseAzureStackProps("proj", "uat")));
    }


    // ── NetworkConstruct ───────────────────────────────────────────────────────

    [Fact]
    public void Network_ModuleSourceIsPinned()
    {
        var stack = PlainStack();
        new NetworkConstruct(stack, "net", new NetworkConstructProps(
            Project: "myapp", Environment: "dev",
            ResourceGroup: "myapp-dev-rg", Location: "eastus",
            AddressSpace: ["10.0.0.0/16"]));

        var mod = Module(Synth(stack), "networking");
        Assert.Contains("ref=v", mod.GetProperty("source").GetString());
    }

    [Fact]
    public void Network_RequiredTagsInjected()
    {
        var stack = PlainStack();
        new NetworkConstruct(stack, "net", new NetworkConstructProps(
            Project: "myapp", Environment: "staging",
            ResourceGroup: "myapp-staging-rg", Location: "eastus",
            AddressSpace: ["10.0.0.0/16"]));

        var tags = Module(Synth(stack), "networking").GetProperty("tags");
        Assert.Equal("terraform", tags.GetProperty("managed_by").GetString());
        Assert.Equal("myapp",     tags.GetProperty("project").GetString());
        Assert.Equal("staging",   tags.GetProperty("environment").GetString());
    }

    [Fact]
    public void Network_SubnetDelegationSerialized()
    {
        var stack = PlainStack();
        new NetworkConstruct(stack, "net", new NetworkConstructProps(
            Project: "myapp", Environment: "dev",
            ResourceGroup: "myapp-dev-rg", Location: "eastus",
            AddressSpace: ["10.0.0.0/16"],
            Subnets: new Dictionary<string, SubnetConfig>
            {
                ["db"] = new SubnetConfig(
                    AddressPrefix: "10.0.8.0/24",
                    Delegation: new SubnetDelegation(
                        Name: "postgres",
                        Service: "Microsoft.DBforPostgreSQL/flexibleServers",
                        Actions: ["Microsoft.Network/virtualNetworks/subnets/join/action"]))
            }));

        var mod = Module(Synth(stack), "networking");
        var service = mod
            .GetProperty("subnets").GetProperty("db")
            .GetProperty("delegation").GetProperty("service")
            .GetString();
        Assert.Equal("Microsoft.DBforPostgreSQL/flexibleServers", service);
    }

    [Fact]
    public void Network_ThrowsOnInvalidEnvironment()
    {
        var stack = PlainStack();
        // NetworkConstruct does not validate environment itself — BaseAzureStack does.
        // This verifies the stack-level guard is sufficient.
        var app = Testing.App();
        Assert.Throws<ArgumentException>(() =>
            new BaseAzureStack(app, "bad",
                new BaseAzureStackProps("proj", "uat")));
    }


    // ── DatabaseConstruct ──────────────────────────────────────────────────────

    [Fact]
    public void Database_ModuleSourceIsPinned()
    {
        var stack = PlainStack();
        new DatabaseConstruct(stack, "db", new DatabaseConstructProps(
            Project: "myapp", Environment: "prod",
            ResourceGroup: "myapp-prod-rg", Location: "eastus",
            SubnetId: "/subscriptions/x/subnets/db",
            DnsZoneId: "/subscriptions/x/privateDnsZones/postgres",
            AdminPassword: "Hunter2!"));

        var src = Module(Synth(stack), "postgres").GetProperty("source").GetString();
        Assert.Contains("ref=v", src);
        Assert.Contains("modules/database/postgres", src);
    }

    [Fact]
    public void Database_HaEnabledSetsZoneRedundant()
    {
        var stack = PlainStack();
        new DatabaseConstruct(stack, "db", new DatabaseConstructProps(
            Project: "myapp", Environment: "prod",
            ResourceGroup: "myapp-prod-rg", Location: "eastus",
            SubnetId: "/subscriptions/x/subnets/db",
            DnsZoneId: "/subscriptions/x/privateDnsZones/postgres",
            AdminPassword: "Hunter2!",
            Config: new PostgresConfig(Databases: ["appdb"], HaEnabled: true)));

        var ha = Module(Synth(stack), "postgres")
            .GetProperty("high_availability_mode").GetString();
        Assert.Equal("ZoneRedundant", ha);
    }

    [Fact]
    public void Database_HaDisabledSetsDisabled()
    {
        var stack = PlainStack();
        new DatabaseConstruct(stack, "db", new DatabaseConstructProps(
            Project: "myapp", Environment: "dev",
            ResourceGroup: "myapp-dev-rg", Location: "eastus",
            SubnetId: "/subscriptions/x/subnets/db",
            DnsZoneId: "/subscriptions/x/privateDnsZones/postgres",
            AdminPassword: "Hunter2!"));

        var ha = Module(Synth(stack), "postgres")
            .GetProperty("high_availability_mode").GetString();
        Assert.Equal("Disabled", ha);
    }

    [Fact]
    public void Database_AdminPasswordPresent()
    {
        var stack = PlainStack();
        new DatabaseConstruct(stack, "db", new DatabaseConstructProps(
            Project: "myapp", Environment: "dev",
            ResourceGroup: "myapp-dev-rg", Location: "eastus",
            SubnetId: "/subscriptions/x/subnets/db",
            DnsZoneId: "/subscriptions/x/privateDnsZones/postgres",
            AdminPassword: "MySecret123!"));

        var mod = Module(Synth(stack), "postgres");
        Assert.True(mod.TryGetProperty("administrator_password", out _),
            "administrator_password should be present in module variables");
    }


    // ── AksConstruct ──────────────────────────────────────────────────────────

    [Fact]
    public void Aks_ModuleSourceIsPinned()
    {
        var stack = PlainStack();
        new AksConstruct(stack, "aks", new AksConstructProps(
            Project: "myapp", Environment: "dev",
            ResourceGroup: "myapp-dev-rg", Location: "eastus",
            SubnetId: "/subscriptions/x/subnets/aks",
            LogWorkspaceId: "/subscriptions/x/workspaces/logs"));

        var src = Module(Synth(stack), "aks").GetProperty("source").GetString();
        Assert.Contains("ref=v", src);
        Assert.Contains("modules/compute/aks", src);
    }

    [Fact]
    public void Aks_AdditionalNodePoolForwarded()
    {
        var stack = PlainStack();
        new AksConstruct(stack, "aks", new AksConstructProps(
            Project: "myapp", Environment: "staging",
            ResourceGroup: "myapp-staging-rg", Location: "eastus",
            SubnetId: "/subscriptions/x/subnets/aks",
            LogWorkspaceId: "/subscriptions/x/workspaces/logs",
            Config: new AksConfig(
                AdditionalNodePools: new Dictionary<string, NodePoolConfig>
                {
                    ["workers"] = new NodePoolConfig(
                        VmSize: "Standard_D8s_v3",
                        EnableAutoScaling: true,
                        MinCount: 2, MaxCount: 10)
                })));

        var pools = Module(Synth(stack), "aks").GetProperty("additional_node_pools");
        Assert.True(pools.TryGetProperty("workers", out var workers));
        Assert.Equal("Standard_D8s_v3", workers.GetProperty("vm_size").GetString());
        Assert.True(workers.GetProperty("enable_auto_scaling").GetBoolean());
    }

    [Fact]
    public void Aks_RequiredTagsInjected()
    {
        var stack = PlainStack();
        new AksConstruct(stack, "aks", new AksConstructProps(
            Project: "myapp", Environment: "prod",
            ResourceGroup: "myapp-prod-rg", Location: "eastus",
            SubnetId: "/subscriptions/x/subnets/aks",
            LogWorkspaceId: "/subscriptions/x/workspaces/logs"));

        var tags = Module(Synth(stack), "aks").GetProperty("tags");
        Assert.Equal("terraform", tags.GetProperty("managed_by").GetString());
        Assert.Equal("prod",      tags.GetProperty("environment").GetString());
    }
}
